geopandas/vectorized.pyx

```diff
@@ -0,0 +1,1029 @@
+
+    handle = get_geos_context_handle()
+
+    for idx in xrange(n):
+        g = L[idx]
+        if g is not None and not (isinstance(g, float) and np.isnan(g)):
+            geos_geom = <np.uintp_t> g.__geom__
+            geom = GEOSGeom_clone_r(handle, <GEOSGeometry *> geos_geom)  # create a copy rather than deal with gc
+            out[idx] = <np.uintp_t> geom
+        else:
+            out[idx] = 0
+
+    return VectorizedGeometry(out)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef serialize(np.ndarray[np.uintp_t, ndim=1, cast=True] geoms):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef uintptr_t geos_geom
+    cdef unsigned int n = geoms.size
+    cdef size_t size
+    cdef GEOSWKBWriter *writer
+    cdef unsigned char* c_string
+    cdef bytes py_string
+
+    cdef np.ndarray[np.uintp_t, ndim=1] sizes = np.empty(n, dtype=np.uintp)
+    cdef bytearray out = bytearray()
+
+    handle = get_geos_context_handle()
+
+    writer = GEOSWKBWriter_create_r(handle)
+
+    for idx in xrange(n):
+        geos_geom = geoms[idx]
+        geom = <GEOSGeometry *> geos_geom
+        if geom:
+            c_string = GEOSWKBWriter_write_r(handle, writer, geom, &size)
+            py_string = c_string[:size]
+            out.extend(py_string)
+            free(c_string)
+            sizes[idx] = size
+        else:
+            sizes[idx] = 0
+
+    GEOSWKBWriter_destroy_r(handle, writer)
+    return out, sizes
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef deserialize(const unsigned char* data, np.ndarray[np.uintp_t, ndim=1, cast=True] sizes):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef uintptr_t geos_geom
+    cdef unsigned int n = sizes.size
+    cdef size_t size
+    cdef GEOSWKBReader *reader
+
+    cdef np.ndarray[np.uintp_t, ndim=1] out = np.empty(n, dtype=np.uintp)
+
+    handle = get_geos_context_handle()
+
+    reader = GEOSWKBReader_create_r(handle)
+
+    with nogil:
+        for idx in xrange(n):
+            size = sizes[idx]
+            if size:
+                geom = GEOSWKBReader_read_r(handle, reader, data, size)
+                out[idx] = <np.uintp_t> geom
+            else:
+                out[idx] = 0
+            data += size
+
+    GEOSWKBReader_destroy_r(handle, reader)
+    return out
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef vec_free(np.ndarray[np.uintp_t, ndim=1, cast=True] geoms):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef uintptr_t geos_geom
+    cdef unsigned int n = geoms.size
+
+    handle = get_geos_context_handle()
+
+    with nogil:
+        for idx in xrange(n):
+            geos_geom = geoms[idx]
+            geom = <GEOSGeometry *> geos_geom
+            if geom:
+                GEOSGeom_destroy_r(handle, geom)
+
+
+class VectorizedGeometry(object):
```

Comment:
Do you have any preference on the name? 
If not, I think I have a slight preference for GeometryArray, but not that important